### DISCUSSION TITLE
architectural pydantic question, also on python in general

### DISCUSSION THREAD

**Discussion by orfisko** [pydantic/pydantic#7033] [General]
Hello, 

I am currently using pydantic to build a configurator to build wooden cabinets. I am already more than two years in this project, and as you can imagine it's quite a thing :). My application calcuates 3d coordinates that are being sent to an angular frontend which uses Three.js for showing the shapes.
 
When defining a panel, I have a lot of stuff to keep in check. The way the panel is connected, which edge is connected, which hardware is being used etc. Hence, there are quite some dependencies between the attributes of my Panel object. 

To give a specific example: I need to specify to the user what kind of connections can be used. For now I have specified an attribute named 'connectionTypeOptions' on the panel which holds enums for the relevant ConnectionType. This works fine for most cases, but not in each and everycase. For some panels, the options that are available depend on the thickness of the board that is being used to make the panel. That board should have a minimum thickness of 18mm before I can allow the user to specify the usage of screws to connect that panel to another panel for example. 
The further I get in this project and the more information is being added, the more dependencies I am getting between the attributes of classes, similar to the one described above. In a java application, one would use setters in order to change these other attributes in case an attribute affects another. So if the board changes to a thickness of 8 mm I could pop the connectionTypeOption of using screws from the connectionTypeOptions attribute by using the board setter. 

Since pydantic v2 we have computed fields, which would allow me to use python properties with setters, but somehow it does not feel right in Python to do it like this. Another option that jumped in my head was hooking in the action of setting an attribute in a pydantic model with an eventlistener function that gets called when an attribute changes. In case I need to adjust other attributes based on that change, I could then match the field that is being updated and trigger the correct logic. I am not sure whether that's a good idea either.

What are your thoughts on this? How do you solve these kind of issues in your applications?

Thanks for any input, 
Hendrik

**Comment by mzamini92** [comment]
There are a few ways to solve the issue you're facing. One option is to use computed fields, as you mentioned. This would allow you to define the connectionTypeOptions field as a computed field that depends on the board field. For example, you could define the connectionTypeOptions field as follows:
```
connectionTypeOptions = computed(
    lambda self: [
        c for c in ConnectionType if c.minBoardThickness <= self.board.thickness
    ]
)
```
This would ensure that the connectionTypeOptions field is always up-to-date with the value of the board field.

Another option is to use an event listener, as you also mentioned. This would allow you to register a function to be called whenever the board field is changed. In your function, you could then update the connectionTypeOptions field to reflect the new value of the board field.

For example, you could define the following event listener:
```
def on_board_change(self, old_board, new_board):
    self.connectionTypeOptions = [
        c for c in ConnectionType if c.minBoardThickness <= new_board.thickness
    ]
```
This would ensure that the connectionTypeOptions field is always updated whenever the board field is changed.

**Comment by orfisko** [comment]
Thanks for your reply. In this specific case I could indeed go with the computed field option. There is however still  a [bug](https://github.com/pydantic/pydantic/issues/6696) (according to me) in pydantic whereby the setter is not being called on a computed field during init. In this case, the type of connections I want to allow is static given the the type of panel I am using, so I can pass them in as a default. As such, they do not need to be overwritten during init (thinking about it, I should even prevent this). But not having this setter being called during init still makes computed_fields no real python properties that also exist during init, which limits their functionality somewhat. 

> Another option is to use an event listener, as you also mentioned. This would allow you to register a function to be called whenever the board field is changed. In your function, you could then update the connectionTypeOptions field to reflect the new value of the board field.
> 
> For example, you could define the following event listener:
> 
> ```
> def on_board_change(self, old_board, new_board):
>     self.connectionTypeOptions = [
>         c for c in ConnectionType if c.minBoardThickness <= new_board.thickness
>     ]
> ```
> 
> This would ensure that the connectionTypeOptions field is always updated whenever the board field is changed.

The bit that I am struggling with is the actual call to this function. How can i detect an update to board? Would I need to overwrite the __set_attr__ dunder method in that case or is there a more refined approach?

**Reply by mzamini92** [reply]
You're right, the computed field option does have some limitations. In your specific case, it would be better to use an event listener.

To detect an update to the board field, you can use the field_changed event. This event is emitted whenever a field on the model is changed. You can register a function to be called whenever this event is emitted. For example, you could do the following:
```
def on_board_change(self, old_board, new_board):
    self.connectionTypeOptions = [
        c for c in ConnectionType if c.minBoardThickness <= new_board.thickness
    ]
```
This would ensure that the on_board_change function is called whenever the board field is changed.

To register your function to be called, you can use the subscribe method. For example, you could do the following:
```
panel = Panel(board=Board(thickness=18))
panel.subscribe(on_board_change)
```
This would register the on_board_change function to be called whenever the board field on the panel object is changed.

**Comment by orfisko** [comment]
> ```
> def on_board_change(self, old_board, new_board):
>     self.connectionTypeOptions = [
>         c for c in ConnectionType if c.minBoardThickness <= new_board.thickness
>     ]
> ```
> 
> This would ensure that the on_board_change function is called whenever the board field is changed.
> 
> To register your function to be called, you can use the subscribe method. For example, you could do the following:
> 
> ```
> panel = Panel(board=Board(thickness=18))
> panel.subscribe(on_board_change)
> ```
> 
> This would register the on_board_change function to be called whenever the board field on the panel object is changed.

I don't see a possibility to subscribe to a basemodel instance in the way you have just explained it. When I test it the subscribe method does not exist on the basemodel instance.
